### PR TITLE
Add support for preserving comments

### PR DESCRIPTION
**Pull request by luxas**
This PR adds support for preserving comments.
It does so first by storing the original file content when decoding in an annotation (if ObjectMeta exists in the object, which is most certainly the case) on the object. Then the object can be freely mutated (conversions, defaulting, whatever).
Finally, when encoded, if the annotation with the orignal file data (base64-encoded) exists, it will use `kyaml` to copy over the comments from the parsed YAML trees from "before" and "after".

You can see the unit test for usage.
cc @bboreham FYI

### CODE REVIEW

File: pkg/serializer/encode.go [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
 )
 
 type EncodingOptions struct {
 	// Use pretty printing when writing to the output. (Default: true)
+	// TODO: Fix that sometimes omitempty fields aren't respected
 	Pretty *bool
+	// Whether to preserve YAML comments internally. Only applicable to ContentTypeYAML framers.
+	// Using any other framer will be silently ignored. Usage of this option also requires setting
+	// the PreserveComments in EncodingOptions, too. (Default: false)
+	PreserveComments *bool
```
Comment (twelho) [inline pkg/serializer/encode.go:23]: Similar to decoder, add a note here that this only works for types implementing `metav1.ObjectMeta`

File: pkg/serializer/encode.go [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
+	"sigs.k8s.io/kustomize/kyaml/yaml"
 )
 
 type EncodingOptions struct {
 	// Use pretty printing when writing to the output. (Default: true)
+	// TODO: Fix that sometimes omitempty fields aren't respected
 	Pretty *bool
+	// Whether to preserve YAML comments internally. Only applicable to ContentTypeYAML framers.
+	// Using any other framer will be silently ignored. Usage of this option also requires setting
+	// the PreserveComments in EncodingOptions, too. (Default: false)
```
Comment (twelho) [inline pkg/serializer/encode.go:22]: EncodingOptions -> DecodingOptions

File: pkg/serializer/encode.go [hunk 3]
```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
 		encoder = serializerInfo.PrettySerializer
 	}
 
+	// Get a version-specific encoder for the specified groupversion
+	versionEncoder := e.codecs.EncoderForVersion(encoder, gv)
+
+	// If the user asked to preserve comments and the contenttype is YAML, try to preserve comments
```
Comment (twelho) [inline pkg/serializer/encode.go:123]: contenttype -> content type

File: pkg/serializer/encode.go [hunk 4]
```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
...
+	metaobj, ok := obj.(metav1.Object)
+	if !ok { // for objects without ObjectMeta, this will fail. allow that failure.
+		// fallback to "normal" encoding
+		return nil, false // fmt.Errorf("couldn't convert to metav1.Object") // TODO: maybe typed errors?
+	}
+
+	// Get annotations and the specific annotation value we're interested in
+	// TODO: This will error if metav1.ObjectMeta is embedded int the object
+	// as a pointer (i.e. *metav1.ObjectMeta) and nil
+	a := metaobj.GetAnnotations()
```
Comment (twelho) [inline pkg/serializer/encode.go:152]: Prevent segfault here as well

File: pkg/serializer/encode.go [hunk 5]
```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
 		encoder = serializerInfo.PrettySerializer
 	}
 
+	// Get a version-specific encoder for the specified groupversion
+	versionEncoder := e.codecs.EncoderForVersion(encoder, gv)
+
+	// If the user asked to preserve comments and the contenttype is YAML, try to preserve comments
+	// If preserving comments was asked for, but the content type isn't YAML, continue with normal operation
+	if *e.opts.PreserveComments && fw.ContentType() == ContentTypeYAML {
+		err, ok := e.tryToPreserveComments(versionEncoder, fw, obj)
```
Comment (twelho) [inline pkg/serializer/encode.go:126]: err -> result, for clarity. We want to conditionally return `nil` from this as well.

File: pkg/serializer/encode.go [hunk 6]
```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
...
+	// Make sure we sanitize the annotation before encoding
+	sanitizeAnnotations(obj)
+
 	// Specialize the encoder for a specific gv and encode the object
-	return e.codecs.EncoderForVersion(encoder, gv).Encode(obj, fw)
+	return versionEncoder.Encode(obj, fw)
+}
+
+func (e *encoder) tryToPreserveComments(versionEncoder runtime.Encoder, fw FrameWriter, obj runtime.Object) (error, bool) {
+	// TODO: Maybe use named returns here?
```
Comment (twelho) [inline pkg/serializer/encode.go:140]: Named returns would make a lot of sense here

File: pkg/serializer/encode.go [hunk 7]
```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
...
+	// Cast the object to a metav1.Object to get access to annotations
+	metaobj, ok := obj.(metav1.Object)
+	if !ok { // if the object doesn't have this ObjectMeta, never mind
+		return
+	}
+
+	// Get annotations and the specific annotation value we're interested in
+	// TODO: This will error if metav1.ObjectMeta is embedded int the object
+	// as a pointer (i.e. *metav1.ObjectMeta) and nil
+	a := metaobj.GetAnnotations()
```
Comment (twelho) [inline pkg/serializer/encode.go:221]: Prevent segfault

File: pkg/serializer/encode.go [hunk 8]
```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
...
+	if a == nil { // if the object doesn't have any annotations, never mind
+		return
+	}
+	_, ok = a[preserveCommentsAnnotation]
```
Comment (twelho) [inline pkg/serializer/encode.go:225]: This code segment is repeated in quite a lot of places, maybe unify into a private function?

File: pkg/serializer/serializer_test.go [hunk 9]
```diff
@@ -159,15 +190,43 @@ func registerNewCRD(scheme *runtime.Scheme) error {
 var _ crdconversion.Hub = &CRDNewVersion{}
 
 type CRDNewVersion struct {
-	runtimetest.ExtensionB
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	OtherString       string `json:"otherString"`
 }
 
 func (t *CRDNewVersion) Hub() {}
```
Comment (twelho) [inline pkg/serializer/serializer_test.go:198]: Could add a comment like `Mark CRDNewVersion has the Hub type to become a conversion target`

File: pkg/serializer/encode.go (hunk 2 above)
Comment (luxas) [reply pkg/serializer/encode.go:15]: this is correct

File: pkg/serializer/encode.go (hunk 5 above)
Comment (luxas) [reply pkg/serializer/encode.go]: rewrote structure

**Pull request merged**